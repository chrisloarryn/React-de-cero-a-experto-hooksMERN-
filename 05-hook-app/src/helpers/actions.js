export const ADD_TODO = 'ADD'
export const DEL_TODO = 'DEL'
export const TOG_TODO = 'TOG'
export const TOGGLE_TODO = 'TOGGLE'
