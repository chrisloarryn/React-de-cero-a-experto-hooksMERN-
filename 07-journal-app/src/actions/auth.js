import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types';

export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {
    setTimeout(() => {
      dispatch(login(123, 'Peter'));
    }, 3500);
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {
    firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      .then((userCred) => {
        console.log(userCred);
      });
  };
};

export const login = (uid, displayName) => ({
  type: types.LOGIN,
  payload: {
    uid,
    displayName
  }
});
