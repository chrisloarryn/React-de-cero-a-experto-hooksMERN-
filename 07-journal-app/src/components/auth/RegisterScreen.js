import React from 'react';
import { Link } from 'react-router-dom';

export const RegisterScreen = () => {
  return (
    <>
      <h3 className='auth__title'>Register</h3>
      <form>
        <input
          type='text'
          name='name'
          placeholder='Name'
          className='auth__input'
          autoComplete='off'
        />
        <input
          type='text'
          name='email'
          placeholder='Email'
          className='auth__input'
          autoComplete='off'
        />
        <input
          type='password'
          name='password'
          placeholder='Password'
          className='auth__input'
        />
        <input
          type='password'
          name='password2'
          placeholder='Confirm password'
          className='auth__input'
        />
        <button className='btn btn-primary btn-block mb-4' type='submit'>
          Register
        </button>
        <hr />
        <Link className='link mt-4' to='/auth/login'>
          Already registered?
        </Link>
      </form>
    </>
  );
};
