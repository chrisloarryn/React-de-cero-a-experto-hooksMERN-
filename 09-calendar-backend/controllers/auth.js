const { response } = require('express')
const bcrypt = require('bcryptjs')
const User = require('./../models/User')
const { generateJWT } = require('./../helpers/jwt')

const createUser = async (req, res = response) => {
  const { email, password } = req.body
  try {
    let user = await User.findOne({ email }).select('-password -__v')
    if (user)
      return res.status(400).json({
        ok: false,
        msg: `Email registered for user: '${user.name}'.`,
        user
      })
    user = new User(req.body)

    // encrypt password
    const salt = bcrypt.genSaltSync()
    user.password = bcrypt.hashSync(password, salt)

    await user.save()

    // generate token
    const token = await generateJWT(user.id, user.name)

    res.status(201).json({
      ok: true,
      uid: user.id,
      name: user.name,
      token
    })
  } catch (err) {
    res.status(500).json({
      ok: false,
      msg: 'Please contact the administrator.'
    })
  }
}

const loginUser = async (req, res = response) => {
  const { email, password } = req.body
  try {
    const user = await User.findOne({ email })
    console.log(user)
    if (!user) {
      return res.status(400).json({
        ok: false,
        msg: `User with email: '${user.name}', doesn\'t exist'.`
      })
    }
    // compare password
    const validPassword = bcrypt.compareSync(password, user.password)
    if (!validPassword) {
      return res.status(400).json({
        ok: false,
        msg: 'Password incorrect.'
      })
    }
    console.log(validPassword)

    // generate token
    const token = await generateJWT(user.id, user.name)

    res.json({
      ok: true,
      uid: user.id,
      name: user.name,
      token
    })
  } catch (err) {
    res.status(500).json({
      ok: false,
      msg: 'Please contact the administrator.'
    })
  }
}

const renewToken = async (req, res = response) => {
  const { name, uid } = req

  // generate token
  const token = await generateJWT(uid, name)
  res.json({
    ok: true,
    token
  })
}

module.exports = { createUser, loginUser, renewToken }
