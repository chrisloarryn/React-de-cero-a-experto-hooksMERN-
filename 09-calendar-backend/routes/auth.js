/**
 * Users Routes / Auth routes
 * host + /api/auth
 */

const { Router } = require('express')
const { check } = require('express-validator')
const { createUser, loginUser, renewToken } = require('../controllers/auth')
const { fieldsValidator } = require('../middlewares/fields-validator')
const { validateJWT } = require('../middlewares/validate-jwt')

const router = Router()

router.post(
  '/new',
  [
    check('name')
      .not()
      .isEmpty()
      .trim()
      .escape()
      .withMessage('Name is required.'),
    check('email').normalizeEmail().isEmail().withMessage('Email is required.'),
    check('password')
      .not()
      .isEmpty()
      .withMessage()
      .isLength({ min: 6, max: 20 })
      .withMessage('Password should be between 6 and 20 characters.'),
    fieldsValidator
  ],
  createUser
)
router.post(
  '/',
  [
    check('email').normalizeEmail().isEmail().withMessage('Email is required.'),
    check('password')
      .not()
      .isEmpty()
      .withMessage()
      .isLength({ min: 6, max: 20 })
      .withMessage('Password should be between 6 and 20 characters.'),
    fieldsValidator
  ],
  loginUser
)
router.get('/renew', validateJWT, renewToken)

module.exports = router
