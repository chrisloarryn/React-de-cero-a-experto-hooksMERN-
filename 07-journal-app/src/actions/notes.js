import Swal from 'sweetalert2';

import { db } from '../firebase/firebase-config';
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from '../helpers/loadNotes';
import { types } from '../types/types';

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    };
    try {
      const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
      dispatch(activeNote(doc.id, newNote));
      dispatch(addNewNote(doc.id, newNote));
    } catch (err) {
      console.log(err);
    }
  };
};

export const activeNote = (id, note) => {
  return {
    type: types.NOTES_ACTIVE,
    payload: {
      id,
      ...note
    }
  };
};

export const addNewNote = (id, note) => ({
  type: types.NOTES_ADD_NEW,
  payload: {
    id,
    ...note
  }
});

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const setNotes = (notes) => ({
  type: types.NOTES_LOAD,
  payload: notes
});

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    if (!note.url) delete note.url;
    const noteToFirestore = { ...note };
    delete noteToFirestore.id;
    try {
      await db
        .doc(`${uid}/journal/notes/${note.id}`)
        .update(noteToFirestore)
        .then(() => {
          dispatch(refreshNote(note.id, noteToFirestore));
          dispatch(activeNote(note.id, noteToFirestore));
          Swal.fire(
            'Saved',
            `Note "${note.title}" has been updated.`,
            'success'
          );
        });
    } catch (err) {
      const errors = () => {
        const title = err.name
          .replace(/([A-Z])/g, ',$1')
          .split(',')
          .filter((f) => f)
          .join(' ');
        return {
          title,
          text: err.toString().includes('No document to update')
            ? 'No document to update.'
            : 'An error has ocurred.'
        };
      };
      errors() &&
        Swal.fire(
          errors().title ? errors().title : 'Error',
          `${errors().text}`,
          'error'
        );
    }
  };
};

export const refreshNote = (id, note) => ({
  type: types.NOTES_UPDATED,
  payload: {
    id,
    note: {
      id,
      ...note
    }
  }
});

export const startUploading = (file) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes;
    Swal.fire({
      title: 'Uploading...',
      text: 'Please wait...',
      showConfirmButton: false,
      showCancelButton: false,
      allowOutsideClick: false,
      willOpen: () => {
        Swal.showLoading();
      }
      /* onBeforeOpen: () => {
        Swal.showLoading();
      } */
    });
    const fileUrl = await fileUpload(file);
    activeNote.url = fileUrl;
    dispatch(startSaveNote(activeNote));
    Swal.close();
  };
};

export const startDeleting = (id) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;
    await db.doc(`${uid}/journal/notes/${id}`).delete();
    dispatch(deleteNote(id));
  };
};

export const deleteNote = (id) => ({
  type: types.NOTES_DELETE,
  payload: id
});

export const noteLogout = () => ({
  type: types.NOTES_LOGOUT_CLEANING
});
