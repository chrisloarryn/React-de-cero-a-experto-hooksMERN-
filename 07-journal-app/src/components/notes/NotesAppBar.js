import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import moment from 'moment';

import { startSaveNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {
  const dispatch = useDispatch();
  const { active } = useSelector((state) => state.notes);
  const handleSave = () => {
    dispatch(startSaveNote(active));
  };
  const handlePictureClick = () => {
    document.querySelector('#fileSelector').click();
  };
  const handleFileChange = (e) => {
    const [file] = e.target.files;
    if (file) {
      dispatch(startUploading(file));
    }
  };
  const activeDate = moment(active.date);
  return (
    <div className='notes__appbar animate__animated animate__fadeIn animate__faster'>
      <span>{activeDate && activeDate.format('MMMM DD, YYYY')}</span>

      <input
        id='fileSelector'
        type='file'
        name='file'
        style={{ display: 'none' }}
        onChange={handleFileChange}
      />
      <div>
        <button className='btn' onClick={handlePictureClick}>
          Picture
        </button>

        <button className='btn' onClick={handleSave}>
          Save
        </button>
      </div>
    </div>
  );
};
