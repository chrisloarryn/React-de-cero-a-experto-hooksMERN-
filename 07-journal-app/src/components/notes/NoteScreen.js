import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { NotesAppBar } from './NotesAppBar';
import { useForm } from '../../hooks/useForm';
import { activeNote, startDeleting } from '../../actions/notes';

export const NoteScreen = () => {
  const { active: note } = useSelector((state) => state.notes);
  const dispatch = useDispatch();
  const [formValues, handleInputChange, reset] = useForm(note);
  const { body, title, id } = formValues;
  const activeId = useRef(note.id);
  const resetValues = useRef(reset);
  useEffect(() => {
    if (note.id !== activeId.current) {
      resetValues.current(note);
      activeId.current = note.id;
    }
  }, [note]);
  useEffect(() => {
    dispatch(activeNote(formValues.id, { ...formValues }));
  }, [formValues, dispatch]);

  const handleDelete = () => {
    dispatch(startDeleting(id));
  };
  return (
    <div className='notes__main-content animate__animated animate__fadeIn animate__faster'>
      <NotesAppBar />

      <div className='notes__content'>
        <input
          type='text'
          name='title'
          placeholder='Some awesome title'
          className='notes__title-input'
          autoComplete='off'
          value={title}
          onChange={handleInputChange}
        />

        <textarea
          name='body'
          placeholder='What happened today'
          className='notes__textarea'
          value={body}
          onChange={handleInputChange}
        ></textarea>

        {note.url && (
          <div className='notes__image animate__animated animate__fadeIn animate__faster'>
            <img
              src={
                note.url
                  ? note.url
                  : 'https://cdn.pixabay.com/photo/2015/04/23/22/00/tree-736885__340.jpg'
              }
              alt='imagen'
            />
          </div>
        )}
      </div>
      <button className='btn btn-danger' onClick={handleDelete}>
        Delete
      </button>
    </div>
  );
};
