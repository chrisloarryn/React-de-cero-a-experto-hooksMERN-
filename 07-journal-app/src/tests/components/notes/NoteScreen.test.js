import React from 'react'
import { Provider } from 'react-redux'
import { mount } from 'enzyme'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import '@testing-library/jest-dom'
import { activeNote } from '../../../actions/notes'
import { startLogout } from '../../../actions/auth'
import { NoteScreen } from '../../../components/notes/NoteScreen'

jest.mock('../../../actions/notes', () => ({
  activeNote: jest.fn()
}))

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState = {
  auth: {
    uid: 1,
    name: 'Christopher'
  },
  ui: {
    loading: false,
    msgError: null
  },
  notes: {
    active: {
      id: 1234,
      title: 'Hello',
      body: 'World',
      date: 0
    },
    notes: []
  }
}

let store = mockStore(initState)
store.dispatch = jest.fn()

const wrapper = mount(
  <Provider store={store}>
    <NoteScreen />
  </Provider>
)

describe('Tests in <NoteScreen />', () => {
  test('should be rendered correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })
  test('should invoke activeNote', () => {
    wrapper.find('input[name="title"]').simulate('change', {
      target: {
        name: 'title',
        value: 'Hello again'
      }
    })
    expect(activeNote).toHaveBeenLastCalledWith(1234, {
      body: 'World',
      title: 'Hello again',
      id: 1234,
      date: 0
    })
  })
})
