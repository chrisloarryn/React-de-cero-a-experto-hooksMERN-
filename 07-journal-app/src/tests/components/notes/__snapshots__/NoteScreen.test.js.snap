// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <NoteScreen /> should be rendered correctly 1`] = `
<div
  className="notes__main-content animate__animated animate__fadeIn animate__faster"
>
  <div
    className="notes__appbar animate__animated animate__fadeIn animate__faster"
  >
    <span>
      December 31, 1969
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="Hello"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today"
      value="World"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
